src/main/java/org/reaktivity/reaktor/internal/memory/DefaultMemoryManager.java

```diff
@@ -0,0 +1,232 @@
+                }
+                if(node.isLeftChild())
+                {
+                    node = node.walkParent();
+                    node = node.walkRightChild();  // TODO optimize
+                }
+                else
+                {
+                    break; // you are root
+                }
+            }
+            else
+            {
+                node = node.walkLeftChild();
+            }
+        }
+
+        if (node.isFull())
+        {
+            return -1;
+        }
+
+        node.fill();
+
+        int memoffset = node.indexInOrder() * node.blockSize();
+        long addressOffset = buffer.addressOffset() + memoffset;
+
+        while (!node.isRoot())
+        {
+            node = node.walkParent();
+            // TODO optimize (can break out quick)
+            if (node.isLeftFull() && node.isRightFull())
+            {
+                node.fill();
+            }
+            node.split();
+        }
+        return addressOffset;
+    }
+
+    public void release0(
+        long offset,
+        int capacity)
+    {
+        offset -= buffer.addressOffset();
+        final int blockSize = calculateBlockSize(capacity);
+        final int order = calculateOrder(blockSize);
+        final int orderSize = 0x01 << order;
+        final int entryIndex = orderSize - 1 + (int) (offset / blockSize);
+        BTreeFlyweight node = btreeRO.wrap(buffer, entryIndex);
+        node.empty();
+        while (!node.isRoot())
+        {
+            node = node.walkParent();
+            if(!node.isRightFullOrSplit() && !node.isLeftFullOrSplit())
+            {
+                node.empty();
+            }
+            else
+            {
+                node.free();
+            }
+        }
+    }
+
+    private BTreeFlyweight root()
+    {
+        return btreeRO.wrap(buffer, 0);
+    }
+
+    private int calculateOrder(int blockSize)
+    {
+        int order = 0;
+        while (largestBlock >> order != blockSize)
+        {
+            order++;
+        }
+        return order;
+    }
+
+    private int calculateBlockSize(
+        int size)
+    {
+        return findNextPositivePowerOfTwo(size);
+    }
+
+    private static int numOfOrders(
+        long largest,
+        int smallest)
+    {
+        int result = 0;
+        while(largest >>> result != smallest)
+        {
+            result++;
+        }
+        return result + 1;
+    }
+
+    private static int bTreeLength(
+        int largestBlockSize,
+        int smallestBlockSize)
+    {
+        int numOfOrders = numOfOrders(largestBlockSize, smallestBlockSize);
+        return (int) Math.ceil(((0x01 << numOfOrders) * BITS_PER_ENTRY) / (BITS_PER_LONG * 1.0));
+    }
+
+    public boolean released()
+    {
+        return root().isFree();
+    }
+
+    private void lock()
+    {
+        while(!buffer.compareAndSetLong(metaDataOffset, 0L, this.id))
+        {
+            Thread.onSpinWait();
+        }
+    }
```

Comment:
Suggest changing this to `tryLock` returning `boolean` indicating success, such that the caller can manage the looping as needed instead.